gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/util/NumberHelper.java

```diff
@@ -312,43 +356,84 @@ private NumberHelper(final BiFunction<Number, Number, Number> add,
             } else if (clazz.equals(BigInteger.class)) {
-                bits = bits < 128 ? 128 : bits;
+                bits = 128;
             } else if (clazz.equals(Float.class)) {
-                bits = bits < 32 ? 32 : bits;
+                bits = Math.max(bits, 32);
                 fp = true;
             } else if (clazz.equals(Double.class)) {
-                bits = bits < 64 ? 64 : bits;
+                bits = Math.max(bits, 64);
                 fp = true;
             } else /*if (clazz.equals(BigDecimal.class))*/ {
-                bits = bits < 128 ? 128 : bits;
+                bits = 128;
                 fp = true;
                 break; // maxed out, no need to check remaining numbers
             }
         }
-        return determineNumberClass(bits, fp);
+        return  new NumberInfo(bits, fp);
+    }
+
+    public static Class<? extends Number> getHighestCommonNumberClass(final Number... numbers) {
+        return getHighestCommonNumberClass(false, numbers);
+    }
+
+    public static Class<? extends Number> getHighestCommonNumberClass(final boolean forceFloatingPoint, final Number... numbers) {
+        NumberInfo numberInfo = getHighestCommonNumberInfo(forceFloatingPoint, numbers);
+        return determineNumberClass(numberInfo.bits, numberInfo.fp);
+    }
+
+    private static Number mathOperationWithPromote(final MathOperation mathOperation, final boolean forceFloatingPoint, final Number a, final Number b) {
+        if (null == a || null == b) return a;
+        NumberInfo numberInfo = getHighestCommonNumberInfo(forceFloatingPoint, a, b);
+        Number result = 0;
+        while (true) {
+            try {
+                result = mathOperation.apply(numberInfo, a, b);
+                if (result.getClass() == BigInteger.class || result.getClass() == BigDecimal.class)
+                {
+                    return result;
+                }
+                if (Double.isInfinite(result.doubleValue()))
+                {
+                    throw new ArithmeticException("Floating point overflow detected");
+                }
+                return result;
+            } catch (ArithmeticException exception) {
+                if (!numberInfo.fp && numberInfo.bits >= 64) {
+                    throw exception;
+                } else if (numberInfo.fp && numberInfo.bits >= 64) {
+                    return result;
+                }
+                numberInfo.bits <<= 1;
+            }
+        }
     }
 
     /**
      * Adds two numbers returning the highest common number class between them.
      *
+     * <p>
+     * This method returns a result using the highest common number class between the two inputs.
+     * If an overflow occurs (either integer or floating-point), the method promotes the precision
+     * by increasing the bit width or switching to floating-point arithmetic, until a suitable type is found.
+     * If no suitable type exists (e.g., for very large integers beyond 128-bit),
```

Comment:
```suggestion
     * by increasing the bit width until a suitable type is found.
     * If no suitable type exists (e.g., for very large integers beyond 64-bit),
```
This PR no longer promotes from integers to floating point, or promotes past `long`. Same applies to the following 3 equivalent comments.